### ISSUE TITLE
Lint masking bits and shift

### ISSUE THREAD

**Issue by tesuji** [rust-lang/rust-clippy#6632]
### What it does

Consider this code snippet:
```rust
pub const fn to_4_be_nibles(op: u16) -> [u8; 4] {
    [
        (op & 0xF000 >> 12) as u8,
        (op & 0x0F00 >> 8) as u8,
        (op & 0x00F0 >> 4) as u8,
        (op & 0x000F) as u8,
    ]
}
``` 

We'd realize that the code is wrong. `>>` has [stronger precedence](https://doc.rust-lang.org/nightly/reference/expressions.html#expression-precedence) than `&` operator.
So we has to use `()` to group `(op & 0F000)` and the like.
I'd like to have a lint about this.

### Categories (optional)

- Kind: clippy::correctness

### Drawbacks

None.

**Comment by llogiq** [comment]
This is *not* a correctness lint. It should not even lint like this, because shifting the mask may well be the intended outcome of the code.

I'm still somewhat sympathetic to introducing brackets here, and the lint should point out the current meaning and ask if that was really intended.

We actually have the `precedence` lint which could be extended to cover this case.
